Mathlib/FieldTheory/PurelyInseparable/Exponent.lean

```diff
@@ -0,0 +1,150 @@
+/-
+Copyright (c) 2025 Michal Staromiejski. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michal Staromiejski
+-/
+import Mathlib.FieldTheory.PurelyInseparable.Basic
+
+/-!
+
+# The exponent of purely inseparable extensions
+
+This file defines the exponent of a purely inseparable extension (if one exists) and
+some related results.
+
+## Main definitions
+
+- `IsPurelyInseparable.HasExponent`: typeclass to assert a purely inseparable field extension
+  `L / K` has an exponent, that is a smallest natural number `e` such that
+  `a ^ ringExpChar K ^ e ∈ K`
+  for all `a ∈ L`.
+- `IsPurelyInseparable.exponent`: the exponent of a purely inseparable field extension.
+- `IsPurelyInseparable.elemExponent`: the exponent of an element of a purely inseparable
+  field extension, that is the smallest natural number `e` such that `a ^ ringExpChar K ^ e ∈ K`.
+  Note: the result is currently meaningful only when `ringExpChar ≠ 1`.
+
+## Tags
+
+purely inseparable
+
+-/
+
+namespace IsPurelyInseparable
+
+variable (K L : Type*)
+
+section Ring
+
+variable [CommRing K] [Ring L] [Algebra K L]
+
+/-- A predicate class on a ring extension saying that there is a natural number `e`
+such that `a ^ ringExpChar K ^ e ∈ K` for all `a ∈ L`. -/
+class HasExponent : Prop where
+  has_exponent' : ∃ e, ∀ a, a ^ ringExpChar K ^ e ∈ (algebraMap K L).range
+
+theorem HasExponent.hasExponent (p : ℕ) [ExpChar K p] [HasExponent K L] :
+    ∃ e, ∀ a, a ^ p ^ e ∈ (algebraMap K L).range :=
+  ringExpChar.eq K p ▸ has_exponent'
+
+open scoped Classical in
+/-- The *exponent* of a purely inseparable extension is the smallest
+natural number `e` such that `a ^ ringExpChar K ^ e ∈ K` for all `a ∈ L`. -/
+noncomputable def exponent [HasExponent K L] : ℕ :=
+  Nat.find ‹HasExponent K L›.has_exponent'
+
+variable {L}
+
+open Classical in
+theorem exponent_def [HasExponent K L] (a : L) :
+    a ^ ringExpChar K ^ exponent K L ∈ (algebraMap K L).range :=
+  Nat.find_spec ‹HasExponent K L›.has_exponent' a
+
+end Ring
+
+section IsDomain
+
+variable [Field K] [Ring L] [IsDomain L] [Algebra K L]
+
+instance [HasExponent K L] : IsPurelyInseparable K L :=
+  let ⟨n, h⟩ := ‹HasExponent K L›.has_exponent'
+  (isPurelyInseparable_iff_pow_mem K (ringExpChar K)).mpr fun x ↦ ⟨n, h x⟩
+
+end IsDomain
+
+section Field
+
+open Polynomial
+
+variable [Field K] [Field L] [Algebra K L] [IsPurelyInseparable K L]
+variable {L}
+
+/-- 'Encoding' of an element `a : L` as the pair `⟨n, y⟩ : ℕ × K` with
+`minpoly K a = X ^ ringExpChar K ^ n - y`. -/
+noncomputable def elemEncode (a : L) : ℕ × K :=
+  let p := ringExpChar K
+  haveI : ExpChar K p := inferInstance
+  Classical.choose <| Prod.exists'.mpr (minpoly_eq_X_pow_sub_C K p a)
```

Comment:
```suggestion
noncomputable def elemEncode (a : L) : ℕ × K :=
  Classical.choose <| Prod.exists'.mpr (minpoly_eq_X_pow_sub_C K (ringExpChar K) a)
```

(untested)

See also the comment I left at Zulip for a possibly better approach.